contracts/TokenDistribute.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.11;
+
+import "./IERC20Metadata.sol";
+import "./ERC20.sol";
+
+contract TokenDistribute {
+    address public contractOwner;
+    mapping(address=>mapping(address=>uint)) private _erc20Balance;  // _erc20Balance[tokenAddress][owner]
+    mapping(address=>uint) private _nativeBalance;
+    mapping(address=>uint) private _distributedErc20;  // Amount of erc20Token distributed in this contract but haven't withdrew by the interns
+    mapping(address=>bool) private _registered;  
+
+    address [] public erc20Tokens;
+
+    modifier onlyOwner() {
+      if (msg.sender != contractOwner) {
+        revert("not the owner of the contract");
+      }    
+      _;
+    }
+
+    modifier nonZeroAddress(address address_) {
+        if (address_ == address(0)) {
+            revert("invalid receiver");
+        }
+        _;
+    }
+
+    constructor ()
+    {
+        contractOwner = msg.sender;  
+    }
+
+    function nativeBalanceOf(address account) external view returns (uint256) {
+        return _nativeBalance[account];
+    }
+
+    function erc20BalanceOf(address tokenAddress, address owner) external view returns (uint256) {
+        return _erc20Balance[tokenAddress][owner];
+    }
+
+    function distributedErc20(address tokenAddress) external view returns (uint256) {
+        return _distributedErc20[tokenAddress];
+    }
+
+    // Anton approves Token to the contract first, then call this function to deposit Token to the contract
+    function depositErc20(address tokenAddress, uint amount) public onlyOwner
+    {
+        ERC20(tokenAddress).transferFrom(msg.sender, address(this), amount);
+        if (!_registered[tokenAddress])
+        {
+            erc20Tokens.push(tokenAddress);
+        }
+    }
+
+    // Distribute 
+    function distributeErc20(address tokenAddress, address to, uint amount) public onlyOwner nonZeroAddress(to)
+    {
+        require(IERC20Metadata(tokenAddress).balanceOf(address(this)) >= _distributedErc20[tokenAddress] + amount, "not enough token to distribute");
+
+        _distributedErc20[tokenAddress] += amount;
+        _erc20Balance[tokenAddress][to] += amount;
+    }
+
+    // Anton transfers ETH directly to interns by calling this function with the amount of ETH
+    function transferNative(address to) public onlyOwner nonZeroAddress(to) payable
+    {
+        require(msg.value>0, "transfer amount = 0");
+
+        _nativeBalance[to] += msg.value;
+    }
```

Comment:
You don't get my point. I can see that if you want to transfer ETH to Alice. You will call this function. then Alice or someone else will call withdrawNative to withdraw ETH to Alice'balance. Why don't you send ETH directly to Alice's balance? In this case you are sending ETH to this contract then this contract will send ETH to Alice, nothing more? 
The intention of this contract, as I understand, is you first send whatever amount of ETH to the contract (but you haven't assign the balance to some addresses). Then you can allocate the balance later, and the designated addresses can withdraw later. This way you can update the _nativeBalance[Alice] many times you want, and is the feature of this contract. If you can not update it, then just send directly to Alice and not use this contract to save gas fee